### DISCUSSION TITLE
Anything else left for the 3.0.0 release?

### DISCUSSION THREAD

**Discussion by isoos** [isoos/postgresql-dart#216] [General]
I've released a new version: https://pub.dev/packages/postgres/versions/3.0.0-beta.1
I'm wondering if anybody has anything else that they wanted to land in the 3.0.0 release?

My last pondering is the `locality` parameter on [`run`](https://pub.dev/documentation/postgres/3.0.0-beta.1/postgres/Pool/run.html) and `runTx`, whether it is named appropriately and should be a separate parameter as-is now. By default I'd leave as-is, but I'm open for any suggestions.

Other than that, I have a few TODOs in the pool implementation for counting the queries (and expiring the connection if the limit is reached), and maybe expose latencies to the endpoint selection method, but those can be added much later, and may not be blocking for 3.0.0.

/cc @simolus3 @osaxma @insinfo @eseidel @slightfoot

**Comment by osaxma** [comment]
#### Error Handling: 
  I haven't checked if anything was implemented about it, but we had a discussion in #119 about error handling and how can the user react when a connection is terminated. 
  
  @simolus3 suggested:  
  > we could expose a `Future<void> get closed` to get notified when the connection is closed, or an additional status getter to query the current connection state.

#### Messages:
I noticed this is marked as an internal but if anyone is using streaming replication protocol, they need to use this API to send their own messages to the server: 
https://github.com/isoos/postgresql-dart/blob/c671d6558652ae653528ca03ba0c69bac69bdc39/lib/src/client_messages.dart#L37-L43

Another organizational note, maybe it'd make sense to move all the messages related `.dart` files into a `messages` folder similar to `auth` and `pool`. 

----

p.s. Thanks for moving this forward with all the recent PRs.

**Reply by isoos** [reply]
@osaxma: I think the error handling is already done, at least a very similar test was added and the code was fixed in https://github.com/isoos/postgresql-dart/blob/master/test/v3_close_test.dart

messages: landed the suggested changes.

**Reply by simolus3** [reply]
Yeah, thanks and congrats for moving this forward!!

AFAIK the error handling bugs are all fixed, but `bool get isOpen` / `Future<void> get closed` getters to query the state of a session aren't part of the current API. I'm neutral on whether we should have them in the first 3.0 release or whether this can be added later. I can open a PR for this soon though.

**Reply by osaxma** [reply]
> I'm neutral on whether we should have them in the first 3.0 release or whether this can be added later. 

I think adding later would be better as the use-cases/requests appear and become clearer.

**Comment by isoos** [comment]
After reading through the recent issues, I think that maybe the custom type converters could also be implemented for the v3 release:
https://github.com/isoos/postgresql-dart/issues/87
https://github.com/isoos/postgresql-dart/issues/75

**Comment by isoos** [comment]
Note: I had trouble adopting the planned custom type converters, mostly because of the v2 API was either incompatible or would have required a serious rewrite. Instead, I'm hiding the internals of type converters with https://github.com/isoos/postgresql-dart/pull/239 and with that we can go ahead with v3, remove the legacy v2 api in the 3.1 release, and then add custom type converters...

**Comment by isoos** [comment]
I've released `beta.2`, which I think is a good base candidate for the v3 release. Please take a look at the API and let me know if something is off or would really change before 3.0.0 comes out: https://pub.dev/documentation/postgres/3.0.0-beta.2/

I am planning to remove/hide or at least annotate `@internal` some of the `TypeRegistry` and maybe a few more public APIs. At this point I'm not looking to add features, only to make sure we allow room to add things easily if needed.

**Reply by isoos** [reply]
Note: I've merged the changes that I had in mind, and don't really see anything that is a blocker for the 3.0 release.

**Reply by simolus3** [reply]
I just had a quick look at the docs as well. `TypeRegistry` still being exported was a little surprising given that it has no public members, but since it's used on public fields in e.g. `ConnectionSettings`, this is probably the easiest way to deal with this until 3.1.

So the current state looks great to me as well.

**Comment by isoos** [comment]
Note: published the 3.0.0 release. Thank you for all the contributions, nudges and I hope we have created a new baseline that will serve all of us well in the future.
